homeassistant/components/tplink_omada/sensor.py

```diff
@@ -0,0 +1,148 @@
+from homeassistant.helpers.typing import StateType
+
+from . import OmadaConfigEntry
+from .const import OmadaDeviceStatus
+from .coordinator import OmadaDevicesCoordinator
+from .entity import OmadaDeviceEntity
+
+# Useful low level status categories, mapped to a more descriptive status.
+DEVICE_STATUS_MAP = {
+    DeviceStatus.PROVISIONING: OmadaDeviceStatus.PENDING,
+    DeviceStatus.CONFIGURING: OmadaDeviceStatus.PENDING,
+    DeviceStatus.UPGRADING: OmadaDeviceStatus.PENDING,
+    DeviceStatus.REBOOTING: OmadaDeviceStatus.PENDING,
+    DeviceStatus.ADOPT_FAILED: OmadaDeviceStatus.ADOPT_FAILED,
+    DeviceStatus.ADOPT_FAILED_WIRELESS: OmadaDeviceStatus.ADOPT_FAILED,
+    DeviceStatus.MANAGED_EXTERNALLY: OmadaDeviceStatus.MANAGED_EXTERNALLY,
+    DeviceStatus.MANAGED_EXTERNALLY_WIRELESS: OmadaDeviceStatus.MANAGED_EXTERNALLY,
+}
+
+# High level status categories, suitable for most device statuses.
+DEVICE_STATUS_CATEGORY_MAP = {
+    DeviceStatusCategory.DISCONNECTED: OmadaDeviceStatus.DISCONNECTED,
+    DeviceStatusCategory.CONNECTED: OmadaDeviceStatus.CONNECTED,
+    DeviceStatusCategory.PENDING: OmadaDeviceStatus.PENDING,
+    DeviceStatusCategory.HEARTBEAT_MISSED: OmadaDeviceStatus.HEARTBEAT_MISSED,
+    DeviceStatusCategory.ISOLATED: OmadaDeviceStatus.ISOLATED,
+}
+
+
+def _map_device_status(device: OmadaListDevice) -> str:
+    """Map the API device status to the best available descriptive device status."""
+    return DEVICE_STATUS_MAP.get(
+        device.status,
+        DEVICE_STATUS_CATEGORY_MAP.get(
+            device.status_category, OmadaDeviceStatus.UNKNOWN
+        ).value,
+    )
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: OmadaConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up sensors."""
+    controller = config_entry.runtime_data
+
+    devices_coordinator = controller.devices_coordinator
+
+    entities: list[OmadaDeviceEntity] = []
+    for device in devices_coordinator.data.values():
+        entities.extend(
+            OmadaDeviceSensor(devices_coordinator, device, desc)
+            for desc in OMADA_DEVICE_SENSORS
+            if desc.exists_func(device)
+        )
+
+    async_add_entities(entities)
+
+
+@dataclass(frozen=True, kw_only=True)
+class OmadaDeviceSensorEntityDescription(SensorEntityDescription):
+    """Entity description for a status derived from an Omada device in the device list."""
+
+    exists_func: Callable[[OmadaListDevice], bool] = lambda _: True
+    update_func: Callable[[OmadaListDevice], StateType]
+
+
+OMADA_DEVICE_SENSORS: list[OmadaDeviceSensorEntityDescription] = [
+    OmadaDeviceSensorEntityDescription(
+        key="device_status",
+        translation_key="device_status",
+        device_class=SensorDeviceClass.ENUM,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        update_func=_map_device_status,
+        options=[v.value for v in OmadaDeviceStatus],
```

Comment:
Unknown should not be a valid value as it looks weird on the automation editor

Reply:
I went for "Error". An alternative might be to just throw if we ever get an unexpected status, but that might affect all of the other sensors attached to the same coordinator too.

Reply:
Oh sorry let me also add (I was about to go to sleep when I reviewed it), if you return None, it's seen as unknown. Hence we don't want unknown twice

Reply:
Ok, I'll drop the error code and return `None` in that case.